src/coreclr/hosts/inc/coreclrhost.h

```diff
@@ -27,7 +27,8 @@ CORECLR_HOSTING_API(coreclr_initialize,
 
 CORECLR_HOSTING_API(coreclr_shutdown,
             void* hostHandle,
-            unsigned int domainId);
+            unsigned int domainId,
```

Comment:
@janvorli I went through the code from coreclr initialization to shutdown. It seems like we need more efforts to address this issue. Therefore I just wrote down the design here before diving to the code. Please correct me if anything is incorrect. 

Current process of executing a managed assembly:
1. In `ExecuteManagedAssembly`, `coreclr_initialize` called to initialize the `hostHandle`.
2. in `coreclr_initialize` , a `CorHost2` object created, who implemented interface `ICLRRuntimeHost2`, and a pointer to `ICLRRuntimeHost2` is assigned to `hostHandle`.
3. In `ExecuteManagedAssembly`, `coreclr_execute_assembly` called with `hostHandle` passed in.
4. In `coreclr_execute_assembly`, `hostHandle` is cast to `ICLRRuntimeHost2`.
5. In `ExecuteManagedAssembly`, `coreclr_shutdown` called with `hostHandle` passed in.
6. In `coreclr_shutdown`, `hostHandle` is cast to `ICLRRuntimeHost2`.

To fix the `Environment.ExitCode` issue while keeping `ICLRRuntimeHost2` not being changed, I'm going to:
1. Create a new COM interface `ICLRRuntimeHost4` which introduces a new `UnloadAppDomain` method with `latchedExitCode` as its last parameter. 
2. Create a `CorHost3` object which implements `ICLRRuntimeHost4`.
3. Create `coreclr_initialize_2`, `coreclr_execute_assembly_2` and `coreclr_shutdown_2` functions. They have same logic to `coreclr_initialize`, `coreclr_execute_assembly` and `coreclr_shutdown` but use `CorHost3` and `ICLRRuntimeHost4` inside.